Grid actions column - WIP

haynesjm42 opened the pull request:
Initial pass at support actions grid columns. #549 

Extended the Column API to accept `rendererParams` to be passed through to the `renderer` or `elementRenderer` (the mechanism for passing the action definitions through to the ActionBar component)

ActionBar component which accepts a collection of action definitions, and simply renders them in a row as minimal buttons. Action definition properties can be a value or a function which returns a value (where it makes sense).

Also made some changes for #575 as part of this

@lbwexler did a quick review with @amcclain of this approach, wanted your feedback on whether this seems like a sounds approach or if something more object-oriented and explicit (ActionColumn, pass actions to the GridModel, etc.) would be preferable.

See https://github.com/exhi/toolbox/commit/035f035a044385770d2e1b9467fbe7c3b7bc94d3 for an action column example.

haynesjm42 commented:
@amcclain / @lbwexler per discussion with Lee, took a stab at making ActionBar into StoreActionBar and functioning similar to StoreContextMenu, and sharing item configs. Also went ahead and made RestGridModel configurable to show the Edit and Delete actions either in the toolbar or in a column.

See https://github.com/exhi/toolbox/commit/c7acfae2daecf6191e485c7b863d6d0bfe1d1085 for updated toolbox example

lbwexler commented on desktop/cmp/contextmenu/ContextMenuItem.js:
```diff
@@ -11,13 +11,13 @@ import {assign} from 'lodash';
     icon;
-    action;
+    actionFn;
     items;
```
John and I discusssed this name change, due to the fact that the overall object is likely to be called "action".  Will need to determine if this is worth the risk and makes sense.

lbwexler commented on desktop/cmp/contextmenu/StoreContextMenu.js:
```diff
@@ -6,7 +6,7 @@
 
 import {isString} from 'lodash';
-import {StoreContextMenuItem} from './StoreContextMenuItem';
+import {StoreActionDefinition} from '@xh/hoist/desktop/cmp/store/StoreActionDefinition';
```
Word 'Definition' seems a bit clunky.  Why not 'StoreAction'?

lbwexler replied:
Or better yet "RecordAction" and "RecordActionBar" etc...

lbwexler commented on cmp/record/RecordAction.js:
```diff
@@ -5,38 +5,44 @@
-     * @param {Object} c - StoreContextMenuItem configuration.
-     * @param {string} c.text - label to be displayed.
+     * @param {Object} c - RecordAction configuration.
+     * @param {string|null} [c.text] - label to be displayed.
```
The 'string|null' seems clunky to me.  If we really want to express nullability, we could use {?string} or {!string}.  Not sure what we would consider the "default" but we would probably want to decide about that as well (probably non-nullable is the implicit default).

see http://usejsdoc.org/tags-type.html

haynesjm42 replied:
Didn't know about the ? and ! qualifiers, that's handy. I think I added the |null because I made the text optional, and then added the default of null in the ctor to be consistent with the other props and intellij complained about it not being null-able.  Looking at it again I don't think it's necessary to default to null, just let it default to undefined and be optional

lbwexler commented on columns/Column.js:
```diff
@@ -41,11 +41,14 @@ export class Column {
      * @param {boolean} [c.movable] - false to prevent user from drag-and-drop re-ordering.
      * @param {Column~rendererFn} [c.renderer] - function to produce a formatted string for each cell.
-     *      Supports HTML as output. Passed both field value and entire row data object.
+     *      Supports HTML as output. Passed the field value as the first parameter, and an object
```
I think you should remove the detailed type info from this text, and rely on the jsdoc definition of Column~rendererFn instead (which is now out of date).

lbwexler commented on columns/Column.js:
```diff
@@ -70,6 +73,8 @@ export class Column {
      *      itself, and are not all guaranteed to be compatible with its usages of Ag-Grid.
      *      @see {@link https://www.ag-grid.com/javascript-grid-column-properties/|AG-Grid docs}
+     * @param {...*} [rest] - additional properties to store on the column, made available to column
+     *      renderers and other callbacks
```
Last clause of this comment confusing, would leave off

lbwexler commented on columns/Column.js:
```diff
@@ -41,11 +41,14 @@ export class Column {
-     *      Supports HTML as output. Passed both field value and entire row data object.
+     *      Supports HTML as output. Passed the field value as the first parameter, and an object
+     *      containing the record, column, and renderer params (destructured) as the second parameter.
      * @param {function} [c.elementRenderer] - elementFactory function to return a React component
```
can we remove the word 'elementFactory' here.  It seems like a loaded term that makes me think of `elemFactory`.
Can just be a "function returning a react component"

lbwexler commented on columns/Column.js:
```diff
@@ -192,13 +200,19 @@ export class Column {
+
+                    return elementRenderer({value, record, agParams, column});
+                }
                 refresh() {return false}
```
If we are going to mess with these, elementRenderer and renderer callbacks really should end up with the same exact signature.  I think I would prefer renderer(val, Object) but would also be ok with renderer(Object).  In any case, we need to choose one and use the JSDoc callback notation to clearly document it.

haynesjm42 replied:
The only problem with renderer(val, Object) is that you then can't just provide an elemFactory for the elementRenderer. Maybe that's fine and we want to make it clear that the elementRenderer callback and elemFactory are not the same thing? 

I don't think we want to change the cellRenderer signature to take an object, @amcclain made a good point in #575 about that

lbwexler replied:
OK -- totally fine with going that way.  I also think its totally fine that you can't provide a raw elemFactory to the renderer -- as you saw from my other comment want to move away from any possible conflation of the two concepts....

lbwexler commented on desktop/cmp/contextmenu/StoreContextMenu.js:
```diff
@@ -40,8 +40,8 @@ export class StoreContextMenu {
         this.gridModel = gridModel;
         this.items = items.map(it => {
             if (isString(it)) return this.parseToken(it);
```
Need to update JSDoc above here for type of items

lbwexler commented on desktop/cmp/record/RecordActionBar.js:
```diff
@@ -0,0 +1,65 @@
+    constructor(props) {
+        super(props);
+
+        this.actions = props.actions.filter(it => it !== null).map(it => new RecordAction(it));
```
This always creates a RecordAction, but prop type claims that application can pass in one 'already built'.

lbwexler commented on desktop/cmp/record/RecordActionBar.js:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    render() {
+        const {actionsShowOnHover = true} = this.props,
```
I might just call this "showOnHover".  Also, add it to the prop types.

lbwexler commented on desktop/cmp/rest/RestGridModel.js:
```diff
@@ -118,26 +165,20 @@ export class RestGridModel {
-                    icon: Icon.edit(),
-                    action: (item, record) => this.editRecord(record),
-                    recordsRequired: 1
+                    ...this.editAction
```
Do these need to be cloned here?   also think Add should be defined like the other two for symmetry.

lbwexler commented on desktop/cmp/rest/RestGridModel.js:
```diff
@@ -57,18 +80,42 @@ export class RestGridModel {
     }) {
         this.actionEnabled = Object.assign(this.actionEnabled, actionEnabled);
         this.actionWarning = Object.assign(this.actionWarning, actionWarning);
+        this.actionColEnabled = actionColEnabled;
```
If we exported the actions separately, would it be trivial matter for the application to simply specify a column of the form `{...actionsCol, actions: [recordEditAction, recordDeleteAction]}`.

Advantage here is that the application has flexibility to place and size the column, but also mix in its own actions in to the same column.

RestGridModel could even, say,  make its actions toolbar a public property, which could be enhanced along the same lines.

This generally suggests a great pattern for any API -- provide Action configs as public exports which can then be fed back to Grids as applications see fit.  Very similar to our column API.
Could be much more extensible than the 'token' based approach we were using.

lbwexler replied:
If we don't want to do my suggested change now (which I totally understand).  I do think we should remove this property 'actionColEnabled' from RestGrid.  I don't think its the right API, and we have no immediate need for it.   We can take some more time to think about the best pattern for exposing actions in this and other APIs.

lbwexler commented:
Looking

lbwexler commented on desktop/cmp/grid/Grid.js:
```diff
@@ -227,12 +224,14 @@ export class Grid extends Component {
 
         // Prepare each item
         const items = menu.items;
         items.forEach(it => {
```
Thinks this change makes sense -- seems more flexible to not tie the action to a selectionModel specifically.
I still wonder whether we would want to make prepareFn and actionFn take a config object to make them more flexible for gaining additional info in the future (e.g. column, cell, gridModel).  I am ok with waiting on that change if we feel we want to move forward with this now.

haynesjm42 replied:
I have the change for passing a config object to prepareFn and actionFn on another branch, along with other changes to refine the RestGrid "actions" api as we had discussed. Agree that those changes don't need to hold up the action col being merged.

lbwexler replied:
Perfect! -- if we can just unwind the intermediary RestGrid API changes in this branch, I would be fine to merge

haynesjm42 commented on desktop/cmp/grid/ColChooserModel.js:
```diff
@@ -52,9 +52,9 @@ export class ColChooserModel {
 
```
Sorry should have called this out and given some context. When testing the action col, I found that it wasn't working with the column chooser and found it was because the action col does not specify a field (nor should it), which led to this change. I would consider this a bug, as the field is an optional property on the column and is not guaranteed to be unique, we shouldn't be using it as a unique identifier for columns.

lbwexler replied:
great! Nice fix

haynesjm42 commented:
Implemented the changes we discussed and made a few tweaks/fixes.

+ Added support for specifying the header and cell css classes on Columns
+ Added a helper function for calculating the action column width, given the number of actions. This references an exported css var. Seems to work well, but have never done anything like this before so not sure if this is something we want to do or if I am following the desired pattern here
+ Changed the show on hover css logic to use `visibility` instead of `display` because I noticed that the icon centering was off when show on hover was turned on, and it was because I was overriding the `display: inline-flex` from the bp button css with `display: unset`
